src/tuner/nccl_ofi_model.c

```diff
@@ -1,123 +1,260 @@
+ * can be computed better once we are able to model the pipelining, but adding
+ * these base latencies seem to pick better switchpoints. Values in µsecs.
+ */
+static const float nccl_base_lat[NCCL_NUM_ALGORITHMS][NCCL_NUM_PROTOCOLS] = {
+	{6.8, 14.0, 0},   /* Tree */
+	{6.6, 14.0, 8.4}, /* Ring */
+	{0, 0, 0},        /* Collnet Direct */
+	{0, 0, 0},        /* Collnet Chain */
+	{0, 0, 0},        /* NVLS */
+	{0, 0, 0}         /* NVLS Tree */
+};
+
+/*
+ * Latency in µsecs. Note, this is currently different from the network
+ * plugin's param for net latency by design. When we merge with the
+ * platform_data values, we will need to do some additional testing on the base
+ * case where a tuner is not loaded to make sure the same defaults make sense
+ * across both paths, and combine the parameters. fixme: shortcomings in the
+ * model forced this param high to get the right answers.
+ */
+#define NCCL_OFI_TUNER_NETWORK_LATENCY            (75)
+
+/*
+ * With EFA, we expect some fixed cost in the device. This parameter was chosen
+ * by hackishly fitting against measurements.
+ */
+#define NCCL_OFI_TUNER_COMPLETION_OVERHEAD        (7.0)
+
+#define NCCL_OFI_TUNER_LL_PROTO_BW_FACTOR         (0.5)
+#define NCCL_OFI_TUNER_LL128_PROTO_BW_FACTOR      (0.9375)
+#define NCCL_OFI_TUNER_SIMPLE_PROTO_BW_FACTOR     (1.0)
+
+static double all_reduce_ring_cost(struct nccl_ofi_tuner_model_dims const* dims,
+				   const int proto,
+				   const size_t nBytes)
 {
-	/*
-	 * Just passing up the NCCL base latencies for now. These costs could be
-	 * computed too, but that can come as a follow up.
-	 */
-	return nccl_base_lat[algo][proto];
+	double base_latency = nccl_base_lat[NCCL_ALGO_RING][proto];
+	const double p2p_latency = nccl_nvlink_lat[NCCL_ALGO_RING][proto];
+
+	const double logranks = log(dims->num_ranks) / log(2);
+	const double logbytes = log(nBytes) / log(2);
+	const double lograt = (logbytes - logranks);
+	if (floor(lograt) < 23) {
+		/* Hack until we understand this better. */
+		return INFINITY;
+	}
+
+	const double steps_per_rotation = dims->num_ranks;
+	const double num_rotations = 2;
+	const double num_steps = (num_rotations * steps_per_rotation) - 1;
+	const double num_steps_internode = (num_rotations * dims->num_nodes) - 1;
+	const double num_steps_intranode = num_steps - num_steps_internode;
+
+	switch (proto) {
+	case NCCL_PROTO_LL: {
+		const double per_proto_bandwidth_overhead = NCCL_OFI_TUNER_LL_PROTO_BW_FACTOR;
+		const double bandwidth = per_proto_bandwidth_overhead *
+					 ((NCCL_OFI_TUNER_INTERNODE_BW * num_steps_internode) +
+					  (NCCL_OFI_TUNER_INTRANODE_BW * num_steps_intranode));
+
+		const double net_latency = NCCL_OFI_TUNER_NETWORK_LATENCY;
+		const double internode_latency = (base_latency + net_latency) * num_steps_internode;
+		const double intranode_latency = p2p_latency * num_steps_intranode;
+		const double latency = internode_latency + intranode_latency;
+
+		return latency + (num_steps * nBytes) / bandwidth;
```

Comment:
This is definitely wrong; there's a `/ n` missing somewhere.  YOu have the `2 * (n - 1)` from num_steps, but then don't divide.  Also, that bw calculation is kind of weird.  I don't understand how this could be compensated for in the bw number values that are precalculated, so *shrug*?

Reply:
Definitely wrong, but until i come to a better understanding of why ring cost doesn't scale as expected around this boundary, and given that ring always performs better than nvlstree below this boundary (plotting ranks against cost with msize fixed), I'd rather keep the cost underestimated in the interest of ensuring that we make the right decisions. Ring /never/ wins above this boundary and always wins below this boundary.